hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,740 @@
+	Value[index] = val;
+	index = GetNext();
+	Sum -= lastv;
+	Sum += val;
+	return lastv;
+}
+double CircularBuffer::GetLast()
+{
+	//return last recorded value
+	return Value[index];
+}
+
+double CircularBuffer::GetSum()
+{
+	return Sum;
+}
+
+LastValue::LastValue(float pdelta)
+{
+	Delta = pdelta;
+}
+double LastValue::Get(int index)
+{
+	return LastValues[index];
+}
+void   LastValue::Put(int index, double value)
+{
+	LastValues[index] = value;
+}
+bool LastValue::AsChanged(int index, double value)
+{
+	return AsChanged(index, value, Delta);
+}
+bool LastValue::AsChanged(int index, double value, double delta)
+{
+	if (fabs(LastValues[index] - value)  > delta)
+	{
+		LastValues[index] = value;
+		return true;
+	}
+	return false;
+}
+
+std::string ToString(double value, const char * format = "%f")
+{
+	char buf[64];
+	snprintf(buf, sizeof(buf) - 2, format, value);
+	return buf;
+}
+
+std::string ToString(int value, const char * format = "%d")
+{
+	char buf[64];
+	snprintf(buf, sizeof(buf) - 2, format, value);
+	return buf;
+}
+
+
+//option management
+void getOption(TOptionMap &Option, const char * optionName, int & value)
+{
+	value = atoi(Option[optionName].c_str());
+}
+void getOption(TOptionMap &Option, const char * optionName, long & value)
+{
+	value = atol(Option[optionName].c_str());
+}
+void getOption(TOptionMap &Option, const char * optionName, double & value)
+{
+	value = atof(Option[optionName].c_str());
+}
+void getOption(TOptionMap &Option, const char * optionName, std::string & value)
+{
+	value = (Option[optionName]);
+}
+
+void setOption(TOptionMap &Option, const char * optionName, int  value)
+{
+	Option[optionName] = ToString(value, "%d");
+}
+void setOption(TOptionMap &Option, const char * optionName, double  value)
+{
+	Option[optionName] = ToString(value, "%4.1f");;
+}
+void setOption(TOptionMap &Option, const char * optionName, const char * value)
+{
+	Option[optionName] = value;
+}
+void setOption(TOptionMap &Option, const char * optionName, std::string & value)
+{
+	Option[optionName] = value;
+}
+
+
+VirtualThermostat * m_VirtualThermostat;
+
+VirtualThermostat::VirtualThermostat(const int ID)
+{
+	m_HwdID = ID;
+	m_VirtualThermostat = this;
+	//thermostat mode string
+	SetAvailableMode(AVALAIBLE_MODE);
+
+}
+
+VirtualThermostat::~VirtualThermostat()
+{
+	m_bIsStarted = false;
+}
+bool VirtualThermostat::StartHardware()
+{
+	RequestStart();
+
+	m_stoprequested = false;
+
+	//Start worker thread
+	//Start worker thread
+	m_thread = std::make_shared<std::thread>(&VirtualThermostat::Do_Work, this);
+	SetThreadNameInt(m_thread->native_handle());
+
+	sOnConnected(this);
+	m_bIsStarted = true;
+	return (m_thread != NULL);
+}
+bool VirtualThermostat::StopHardware()
+{
+	m_stoprequested = true;
+	{
+		RequestStop();
+		m_thread->join();
+		m_thread.reset();
+	}
+	m_bIsStarted = false;
+	return true;
+}
+void VirtualThermostat::Do_Work()
+{
+	int sec_counter = 0;
+	_log.Log(LOG_STATUS, "VirtualThermostat: Worker started...");
```

Comment:
you should not call the _log functions, please use **Log** directly (this will also print the hardware name during logging and things can be disabled (info/status/error)

Reply:
ok. done also for Debug.